### PR TITLE
Api v2 files

### PR DESCRIPTION
**Pull request by sirambd**
- [x] migrate file actions
 - [x] migrate file activities
 - [x] migrate categories
 - [x] migrate comment
 - [x] migrate Dropbox
 - [x] migrate favorite
 - [x] migrate files
 - [x] migrate search
 - [x] migrate sharelink
 - [x] migrate trash
 - [x] migrate root directories

### CODE REVIEW

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 1]
```diff
@@ -594,7 +600,7 @@ class ApiRepositoryTest : KDriveTest() {
 
             // Make sure the dropbox has been updated
             with(getDropBox(testFolder)) {
-                assertFalse(data?.emailWhenFinished ?: true, "Email when finished should be false")
+                assertFalse(data?.hasNotification ?: true, "Email when finished should be false")
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:603]: Same as previous comment.

File: app/src/main/java/com/infomaniak/drive/data/models/File.kt [hunk 2]
```diff
@@ -53,68 +55,70 @@ import java.util.*
 @Parcelize
 open class File(
     @PrimaryKey var id: Int = 0,
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/data/models/File.kt:57]:
```suggestion
    @PrimaryKey
    var id: Int = 0,
```

File: app/src/main/java/com/infomaniak/drive/data/models/ShareableItems.kt [hunk 3]
```diff
@@ -17,12 +17,20 @@
...
-@Parcelize
 data class ShareableItems(
-    val invitations: ArrayList<Invitation> = ArrayList(),
-    val teams: ArrayList<Team> = ArrayList(),
-    val users: ArrayList<DriveUser> = ArrayList(),
-) : Parcelable
\ No newline at end of file
+    val emails: ArrayList<FeedbackAccessResource<String, Invitation>> = ArrayList(),
+    val teams: ArrayList<FeedbackAccessResource<Int, Team>> = ArrayList(),
+    val users: ArrayList<FeedbackAccessResource<Int, DriveUser>> = ArrayList(),
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/data/models/ShareableItems.kt:23]: Replace with `MutableList`

File: app/src/main/java/com/infomaniak/drive/ui/addFiles/CreateFolderFragment.kt [hunk 4]
```diff
@@ -73,7 +73,7 @@ open class CreateFolderFragment : Fragment() {
         }
     }
 
-    protected fun canInherit(userList: ArrayList<DriveUser>, teamList: ArrayList<Team>): Boolean {
+    protected fun canInherit(userList: ArrayList<UserFileAccess>, teamList: ArrayList<Team>): Boolean {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/addFiles/CreateFolderFragment.kt:76]: Replace with `MutableList`.

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileDetails/FileDetailsViewModel.kt [hunk 5]
```diff
@@ -67,11 +67,11 @@ class FileDetailsViewModel : ViewModel() {
...
-        emit(ApiRepository.postFileCommentLike(file, fileComment.id))
+        emit(ApiRepository.postLikeComment(file, fileComment.id))
     }
 
     fun postUnlike(file: File, fileComment: FileComment) = liveData(Dispatchers.IO) {
-        emit(ApiRepository.postFileCommentUnlike(file, fileComment.id))
+        emit(ApiRepository.postUnlikeComment(file, fileComment.id))
     }
 
     private suspend fun <T> LiveDataScope<ApiResponse<ArrayList<T>>?>.manageRecursiveApiResponse(
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/fileDetails/FileDetailsViewModel.kt:77]: Replace with `MutableList`.

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/AvailableShareableItemsAdapter.kt [hunk 6]
```diff
@@ -44,9 +44,8 @@ import kotlinx.android.synthetic.main.item_user.view.*
 class AvailableShareableItemsAdapter(
     context: Context,
     private var itemList: ArrayList<Shareable>,
-    var notShareableUserIds: ArrayList<Int> = arrayListOf(),
+    var notShareableIds: ArrayList<Int> = arrayListOf(),
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/AvailableShareableItemsAdapter.kt:47]: Replace all `ArrayList` in this file by `MutableList`.

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareAddUserDialog.kt [hunk 7]
```diff
@@ -83,9 +83,8 @@ class FileShareAddUserDialog : FullScreenBottomSheetDialog() {
         availableUsersAdapter = userAutoCompleteTextView.setupAvailableShareableItems(
             context = requireContext(),
             itemList = fileShareViewModel.availableShareableItems.value ?: AccountUtils.getCurrentDrive().getDriveUsers(),
-            notShareableUserIds = navigationArgs.notShareableUserIds.toMutableList() as ArrayList<Int>,
+            notShareableIds = navigationArgs.notShareableIds.toMutableList() as ArrayList<Int>,
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareAddUserDialog.kt:86]: Replace all `ArrayList` in this file by `MutableList`.

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareDetailsFragment.kt [hunk 8]
```diff
@@ -112,27 +112,35 @@ class FileShareDetailsFragment : Fragment() {
...
                 share.teams.sort()
                 availableShareableItemsAdapter.apply {
                     fileShareViewModel.availableShareableItems.value?.let { setAll(it) }
-                    notShareableUserIds = ArrayList(share.users.map { it.id } + share.invitations.map { it.userId })
+
+                    val userIds = share.users.map { it.id } +
+                            share.invitations.mapNotNull { it.user?.id } +
+                            share.teams.map { team -> team.id }
+
+                    notShareableIds = ArrayList(userIds)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareDetailsFragment.kt:132]: Replace all `ArrayList` in this file by `MutableList`.

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/PermissionsAdapter.kt [hunk 9]
```diff
@@ -45,7 +45,7 @@ class PermissionsAdapter(
     var selectionPosition: Int? = null,
     private var currentUser: User? = null,
     private var isExternalUser: Boolean = false,
-    private var sharedUsers: ArrayList<DriveUser> = ArrayList(),
+    private var sharedUsers: ArrayList<UserFileAccess> = ArrayList(),
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/PermissionsAdapter.kt:48]: Replace all `ArrayList` in this file by `MutableList`.

File: app/src/androidTest/java/com/infomaniak/drive/FileControllerTest.kt [hunk 10]
```diff
@@ -148,7 +148,7 @@ class FileControllerTest : KDriveTest() {
             "permission" to Shareable.ShareablePermission.READ,
             "lang" to Locale.getDefault().language,
         )
-        assertApiResponseData(postFileShare(file, body))
+        assertApiResponseData(addMultiAccess(file, body))
 
         // Get remote files
         val remoteFiles = arrayListOf<File>()
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/FileControllerTest.kt:154]: Replace with `MutableList`.

File: app/src/main/java/com/infomaniak/drive/data/models/FileComment.kt [hunk 11]
```diff
@@ -26,10 +26,17 @@ data class FileComment(
     var body: String,
     var liked: Boolean,
     var likes: ArrayList<DriveUser>?,
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/data/models/FileComment.kt:28]: Replace with `MutableList`.

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 12]
```diff
@@ -563,15 +569,15 @@ class ApiRepositoryTest : KDriveTest() {
 
             // Add a dropBox with default body and get its id
             val dropboxId = createDropBoxForTest(testFolder, maxSize).let {
-                assertTrue(it.emailWhenFinished, "Email when finished must be true")
+                assertTrue(it.hasNotification, "Email when finished must be true")
```
[the remark this answers is not in the document]
Comment (sirambd) [reply app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:572]: We are in a migration branch, and the tests will be rewritten on another PR

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt (hunk 1 above)
Comment (sirambd) [reply app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:603]: We are in a migration branch, and the tests will be rewritten on another PR

File: app/src/main/java/com/infomaniak/drive/ui/menu/TrashViewModel.kt [hunk 13]
```diff
@@ -34,20 +34,20 @@ class TrashViewModel : ViewModel() {
...
                 if (apiResponse.isSuccess()) {
                     val data = apiResponse.data
                     when {
                         data == null -> Unit
-                        data.children.size < ApiRepository.PER_PAGE -> emit(
+                        data.size < ApiRepository.PER_PAGE -> emit(
                             FileListFragment.FolderFilesResult(
                                 parentFolder = file,
-                                files = ArrayList(data.children),
+                                files = ArrayList(data),
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/menu/TrashViewModel.kt:50]: Replace all `ArrayList` in this file by `MutableList`.

File: app/src/main/java/com/infomaniak/drive/utils/Extensions.kt [hunk 14]
```diff
@@ -258,18 +258,16 @@ fun String.isEmail(): Boolean = Patterns.EMAIL_ADDRESS.matcher(this).matches()
 fun MaterialAutoCompleteTextView.setupAvailableShareableItems(
     context: Context,
     itemList: List<Shareable>,
-    notShareableUserIds: ArrayList<Int> = arrayListOf(),
+    notShareableIds: ArrayList<Int> = arrayListOf(),
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/utils/Extensions.kt:261]: Replace all `ArrayList` in this file by `MutableList`.

File: app/src/main/java/com/infomaniak/drive/data/models/Invitation.kt [hunk 15]
```diff
@@ -24,14 +24,11 @@ import kotlinx.android.parcel.Parcelize
...
     val status: String = "",
-    val avatar: String = "",
-    override var permission: String = "",
-    @SerializedName("user_id") val userId: Int = -1,
-    @SerializedName("invit_drive") val invitDrive: Boolean = false,
-    @SerializedName("display_name") val displayName: String? = null,
-    @SerializedName("invit_drive_id") val invitDriveId: Int = -1,
+    val email: String = "",
+    val user: DriveUser? = null,
+    @SerializedName("invitation_drive_id") val invitationDriveId: Int = -1,
```
[the remark this answers is not in the document]
Comment (sirambd) [reply app/src/main/java/com/infomaniak/drive/data/models/Invitation.kt:33]: Same that previously

File: app/src/main/java/com/infomaniak/drive/data/cache/FileController.kt [hunk 16]
```diff
@@ -52,21 +47,31 @@ object FileController {
...
+    private const val MIN_VERSION_CODE = 4_02_000_00
 
+    const val ROOT_ID = Utils.ROOT_ID
     const val FAVORITES_FILE_ID = -1
     const val MY_SHARES_FILE_ID = -2
     const val RECENT_CHANGES_FILE_ID = -4
     private const val PICTURES_FILE_ID = -3
 
+    private val ROOT_FILE get() = File(ROOT_ID, name = "Root", driveId = AccountUtils.currentDriveId)
     private val FAVORITES_FILE = File(FAVORITES_FILE_ID, name = "Favoris")
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/data/cache/FileController.kt:59]:
```suggestion
    private val FAVORITES_FILE = File(FAVORITES_FILE_ID, name = "Favorites")
```

File: app/src/main/java/com/infomaniak/drive/data/models/FileCategory.kt [hunk 17]
```diff
@@ -24,13 +24,14 @@ import java.util.*
...
+    @SerializedName("user_validation")
+    var userValidation: String = "", // State of user validation after auto assignment from AI
+    @SerializedName("is_generated_by_ai")
+    var isGeneratedByAI: Boolean = false,
     @SerializedName("user_id")
     var userId: Int? = null,
-    @SerializedName("added_to_file_at")
-    var addedToFileAt: Date = Date(),
+    @SerializedName("added_at")
+    var addedAt: Date = Date(), //Date when the category was added to file
```
[the remark this answers is not in the document]
Comment (sirambd) [reply app/src/main/java/com/infomaniak/drive/data/models/FileCategory.kt:36]: 🤔 why not

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareViewModel.kt [hunk 18]
```diff
@@ -46,28 +46,18 @@ class FileShareViewModel : ViewModel() {
...
-            "block_comments" to shareLink.blockComments,
-            "block_information" to shareLink.blockInformation,
-        )
+        val shareLinkSettings = shareLink.ShareLinkSettings()
 
-        if (AccountUtils.getCurrentDrive()?.pack != Drive.DrivePack.FREE.value) {
-            body["valid_until"] = shareLink.validUntil?.time?.let { it / 1_000L } ?: ""
-        }
+        shareLinkSettings.validUntil =
+            if (AccountUtils.getCurrentDrive()?.pack != Drive.DrivePack.FREE.value) shareLink.validUntil else null
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareViewModel.kt:56]:
```suggestion
        shareLinkSettings.validUntil = if (AccountUtils.getCurrentDrive()?.pack == Drive.DrivePack.FREE.value) {
            null
        } else {
            shareLink.validUntil
        }
```

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/PermissionsAdapter.kt [hunk 19]
```diff
@@ -113,10 +113,10 @@ class PermissionsAdapter(
     private fun View.setupInheritPermissionUi() {
         if (sharedUsers.isNotEmpty()) {
 
-            sharedUsers.firstOrNull()?.let { firstUser -> mainIcon.loadAvatar(firstUser) }
+            sharedUsers.firstOrNull()?.let { firstUser -> firstUser.user?.let { mainIcon.loadAvatar(it) } }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/PermissionsAdapter.kt:116]:
```suggestion
            sharedUsers.firstOrNull()?.let { firstUser -> firstUser.user?.let(mainIcon::loadAvatar) }
```

File: app/src/main/java/com/infomaniak/drive/data/cache/FileMigration.kt [hunk 20]
```diff
@@ -127,6 +131,154 @@ class FileMigration : RealmMigration {
...
+            // Set embedded objects
+            schema.get(ShareLinkCapabilities::class.java.simpleName)?.isEmbedded = true
+            schema.get(ShareLink::class.java.simpleName)?.isEmbedded = true
+            schema.get(DropBoxValidity::class.java.simpleName)?.isEmbedded = true
+            schema.get(DropBoxSize::class.java.simpleName)?.isEmbedded = true
+            schema.get(DropBoxCapabilities::class.java.simpleName)?.isEmbedded = true
+            schema.get(DropBox::class.java.simpleName)?.isEmbedded = true
+            schema.get(FileVersion::class.java.simpleName)?.isEmbedded = true
+            schema.get(FileConversion::class.java.simpleName)?.isEmbedded = true
+        }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/data/cache/FileMigration.kt:280]:
```suggestion
            oldVersionTemp++
        }
```

File: app/src/main/java/com/infomaniak/drive/data/cache/FileMigration.kt [hunk 21]
```diff
@@ -127,6 +131,154 @@ class FileMigration : RealmMigration {
...
+        //region Migrated to version 4:
+        // - Migrate (File) table to api v2
+        // - Migrate (Right) table to api v2
+        // - Migrate (FileActivity) table to api v2
+        // - Added new field (Dropbox) in File table
+        // - Added new field (FileConversion) in File table
+        // - Added new field (FileVersion) in File table
+        // - Added new field (ShareLink) in File table
+
+        if (oldVersionTemp == 3L) {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/data/cache/FileMigration.kt:144]:
```suggestion
        if (oldVersionTemp == 3L) {
```

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareAddUserDialog.kt [hunk 22]
```diff
@@ -219,9 +211,11 @@ class FileShareAddUserDialog : FullScreenBottomSheetDialog() {
         fileShareViewModel.currentFile.value?.let { file ->
             val body = mutableMapOf(
                 "emails" to ArrayList(selectedItems.invitations.map { it.email }),
-                "user_ids" to ArrayList(selectedItems.users.map { it.id }),
+                "user_ids" to ArrayList(selectedItems.driveUsers.map { it.id }),
                 "team_ids" to ArrayList(selectedItems.teams.map { it.id }),
-                "permission" to newPermission
+                "right" to newPermission,
+                "lang" to Locale.getDefault().language,
```
Comment (JorisBodin) [inline app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareAddUserDialog.kt:217]: And if it is a language not supported by the server, there is no error return?

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareViewModel.kt (hunk 18 above)
Comment (sirambd) [reply app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareViewModel.kt:56]: I don't like this

File: app/src/main/java/com/infomaniak/drive/data/cache/FileMigration.kt (hunk 20 above)
Comment (sirambd) [reply app/src/main/java/com/infomaniak/drive/data/cache/FileMigration.kt:279]: I think it's better not to put it, so it's at the next migration that you have to add it, so you get rid of the warnings

File: app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareViewModel.kt (hunk 18 above)
Comment (sirambd) [reply app/src/main/java/com/infomaniak/drive/ui/fileList/fileShare/FileShareViewModel.kt]: I have it properly but differently

**Pull request merged**